python/dgl/distributed/kvstore.py

```diff
@@ -0,0 +1,1008 @@
+        return res
+
+KVSTORE_PUSH = 901232
+
+class PushRequest(rpc.Response):
+    """Send ID tensor and data tensor to server and update kvstore's data.
+
+    This request has no response.
+
+    Parameters
+    ----------
+    name : str
+        data name
+    id_tensor : tensor
+        a vector storing the data ID
+    data_tensor : tensor
+        a tensor with the same row size of data ID
+    """
+    def __init__(self, name, id_tensor, data_tensor):
+        self.name = name
+        self.id_tensor = id_tensor
+        self.data_tensor = data_tensor
+
+    def __getstate__(self):
+        return self.name, self.id_tensor, self.data_tensor
+
+    def __setstate__(self, state):
+        self.name, self.id_tensor, self.data_tensor = state
+
+    def process_request(self, server_state):
+        kv = server_state.kv_store
+        local_id = kv.part_policy[self.name].to_local(self.id_tensor)
+        kv.push_handler(kv.data_store, self.name, local_id, self.data_tensor)
+        return None
+
+INIT_DATA = 901233
+INIT_MSG = 'Init'
+
+class InitDataResponse(rpc.Response):
+    """Send a confirmation response (just a short string message) of
+    InitDataRequest to client.
+
+    Parameters
+    ----------
+    msg : string
+        string message
+    """
+    def __init__(self, msg):
+        self.msg = msg
+
+    def __getstate__(self):
+        return self.msg
+
+    def __setstate__(self, state):
+        self.msg = state
+
+class InitDataRequest(rpc.Request):
+    """Send meta data to server and init data tensor
+    on server using UDF init function.
+
+    Parameters
+    ----------
+    name : str
+        data name
+    shape : tuple
+        data shape
+    dtype : str
+        data type string, e.g., 'int64', 'float32', etc.
+    policy_str : str
+        partition-policy string, e.g., 'edge' or 'node'.
+    init_func : function
+        UDF init function.
+    """
+    def __init__(self, name, shape, dtype, policy_str, init_func):
+        self.name = name
+        self.shape = shape
+        self.dtype = dtype
+        self.policy_str = policy_str
+        self.init_func = init_func
+
+    def __getstate__(self):
+        return self.name, self.shape, self.dtype, self.policy_str, self.init_func
+
+    def __setstate__(self, state):
+        self.name, self.shape, self.dtype, self.policy_str, self.init_func = state
+
+    def process_request(self, server_state):
+        kv = server_state.kv_store
+        dtype = F.data_type_dict[self.dtype]
+        if kv.is_backup_server() == False:
+            data_tensor = self.init_func(self.shape, dtype)
+            kv.init_data(name=self.name, data_tensor=data_tensor)
+        else: # backup server will read data from shared-memory
+            kv.init_data(name=self.name)
+        # Find the same partition policy from exsiting plolicy list.
+        for _, policy in kv.part_policy.items():
+            if policy.policy_str == self.policy_str:
```

Comment:
why not make the key of the dict the policy name?

Reply:
The key is the data's name. We may have many data using the same policy.